### DISCUSSION TITLE
Environment variables with adapter-node not completely working.

### DISCUSSION THREAD

**Discussion by LGmatrix13** [sveltejs/kit#11502] [Q&A] [answered]
Hi everyone!

I'm using the adapter node to deploy sveltekit on node. It works well, but public (prefixed with PUBLIC_) are rendering in +page.svelte(s) as undefined. envs on +server.ts load fine. I did follow the docs and I am using dotenv: https://kit.svelte.dev/docs/adapter-node#environment-variables. Has anyone had a similar issue? If it helps, I am deploying with the help of [Coolify](https://coolify.io/), which using docker. 

Thank you! :)

**Comment by LGmatrix13** [comment, marked as answer]
Looks like you need to import envs one by one like so:
```
import { PUBLIC_SOME_ENV } from "$env/static/public"
```
Instead of:
```
import { env } from '$env/dynamic/public';

env.PUBLIC_SOME_ENV
```

@Rich-Harris this doesn't seem to be in the adapter-node docs under [environment variables](https://kit.svelte.dev/docs/adapter-node#environment-variables)? If this is the expected result, should I create a PR to add the additional detail to the docs?

**Comment by andersoncardoso** [comment]
I'm having the same issue, my "PUBLIC_" env vars are present on `$env/static/public` but not on `$env/dynamic/public`.  Although they are defined inside a .env file, I tried also exposing then on the env regularly (export PUBLIC_....), and still got the same result.
My project uses the node adapter, and it's runnning on render.

**Comment by exx-trost** [comment]
@LGmatrix13  I ran into the same issue. I want to use dynamic public variables (= not available at build time), so your solution is not applicable, right? At least if I try it with $env/dynamic/public, it doesn't compile obviously.

This is blocking us from using adapter-node unfortunately.

**Comment by andersoncardoso** [comment]
Are you running the node server directly with `node build/index.js` ? 
I found this wouldn't work (now looking back, seems obvious since we're not exposing any variables this way). 
So I had to use one of these options: 
```
vite preview

# OR 
export $(cat .env) && node build/index.js 
# note: my env vars would break if the value was inside double quotes

# OR (good old trusty dotenv)
node -r dotenv/config build/index.js
```

then you can use on your backend with: 
```
import { env } from "$env/dynamic/private";
env["MY_DYNAMIC_PRIVATE_VAR"]
```

and on the frontend with: 
```
import { env } from "$env/dynamic/public";
console.log("public app host", env["PUBLIC_APP_HOST"]);
```

**Comment by nmfrankel** [comment]
I'm running into this issue for the second on **two** production builds. 

We are deploying our projects using docker so we are limited in what fixes we can apply. The bigger problem is our build **fails**.

```
Error: Cannot read values from $env/dynamic/private while prerendering (attempted to read env.DATABASE_URL). Use $env/static/private instead

ELIFECYCLE  Command failed with exit code 1.
```

**Comment by therealPaulPlay** [comment]
Just spent a long time debugging this in a remote environment – I was using dynamic `.env` variables together with the adapter-node. This should be documented :-(

`import * as env from '$env/static/private';` this works just fine, `import { env } from '$env/dynamic/private';` does not with the node adapter
